### PR TITLE
[CURA-10722] Add 'additional settings providers' (Engine Plugins)

### PR DESCRIPTION
**Pull request by rburema**
Plugins should now be able to add settings. (But not append enum-values to existing ones, that's another ticket.)

- (De)serialization works, including load/save.
- Keys are tagged/'namespaced' with provider type (just a plugin for now), unique id, version.
- Will be sent to engine (except you have to demangle the tagged/namespaced name a bit).
- Setting-formulas work (except for taking values of another plugin).
- Can add to an existing category, or create a new one.

See (now) also the accompanying front-end PR here: https://github.com/Ultimaker/Cura/pull/16305

### CODE REVIEW

File: UM/Settings/DefinitionContainer.py [hunk 1]
```diff
@@ -332,16 +332,52 @@ def deserialize(self, serialized: str, file_name: Optional[str] = None) -> str:
...
-            definition = SettingDefinition(key, self, None, self._i18n_catalog)
+        self._deserializeDefinitions(parsed["settings"])
+        return serialized
+
+    def _deserializeDefinitions(self, settings_dict: Dict[str, Any], force_category: Optional[str] = None) -> None:
+
+        # When there is a forced category (= parent) present, find the category parent, create it if it doesn't exist.
+        category_parent = None
+        if force_category:
+            category_parent = self.findDefinitions(key=force_category)
```
Comment (jellespijker) [inline UM/Settings/DefinitionContainer.py:343]:
```suggestion
            category_parent = self.findDefinitions(key = force_category)
```

File: UM/Settings/DefinitionContainer.py [hunk 2]
```diff
@@ -332,16 +332,52 @@ def deserialize(self, serialized: str, file_name: Optional[str] = None) -> str:
...
             self._updateRelations(definition)
 
-        return serialized
+    def appendAdditionalSettingDefinitions(self, additional_settings: Dict[str, Dict[str, Any]]) -> None:
+        try:
+            merge_with_existing_categories = {}
+            create_new_categories = {}
+
+            for category, values in additional_settings.items():
+                if len(self.findDefinitions(key=category)) > 0:
```
Comment (jellespijker) [inline UM/Settings/DefinitionContainer.py:365]:
```suggestion
                if len(self.findDefinitions(key = category)) > 0:
```

**Review by jellespijker — CHANGES_REQUESTED** [review]
LGTM; As we discussed during the stand-up we will add functionality to the plugin system as a whole (both front-end and curaengine_plugins) which will allow to add settings by simply specify a path to a `json` file. Which will have setting entries such as they are defined in `fdmprinter.def.json`

**Comment by jellespijker** [comment]
Also pinging @fieldOfView because I think he also could have some valuable insights on this PR and at the very least FYI

**Comment by fieldOfView** [comment]
> Also pinging @fieldOfView 

'sup?

Hmm, I don't quite get why this would be necessary. Many of my plugins are adding settings that eventually also get sent to CuraEngine (which happily ignores them), then to be used in postprocessing. It is all a bit hacky (private method access), but has been working very well for years.

**Comment by rburema** [comment]
@fieldOfView -- If you don't see the need to change for your own plugins, that's fine. We're basically using the same mechanism anyway, so we won't clash.

Since (other) people are (we hope) going to be writing engine-plugins (also), there is additional need to make a proper (non-hacky) way to define additional settings for engine plugins. -- We may trust _you_ with our internals, but that's not necessarily the case for _every_ developer out there.

Also to us these settings do not feel quite the same as plugin preferences, and we'd like the possibility of defining values for them in profiles (for ex.; a 'weird printer' plugin that comes with an engine part and a bunch of profiles for example).

Additionally, different plugins might have the same setting-definition name added. We tag/'sort-of-namespace' the settings internally so that they at least don't destabilize the system. A similar story for different versions of the same plugin that might have tweaked their setting-definition meanings.

File: UM/Settings/AdditionalSettingDefinitionAppender.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                # Replace key-name 'heading'.
+                key = key_map.get(key, key)
+                value = _doReplace(value)
+            result[key] = value
+        return result
+    return _doReplace(settings)
+
+
+class AdditionalSettingDefinitionsAppender(PluginObject):
+
```
Comment (nallath) [inline UM/Settings/AdditionalSettingDefinitionAppender.py:77]: What is an additional settings defintions appender? When should one use it? (aka: I think that this needs documentation)

File: UM/Settings/DefinitionContainer.py [hunk 4]
```diff
@@ -332,16 +332,52 @@ def deserialize(self, serialized: str, file_name: Optional[str] = None) -> str:
...
+                # Forced category; these are then children of that category, instead of full categories on their own.
+                category_parent.children.append(definition)
+            else:
+                self._definitions.append(definition)
 
         for definition in self._definitions:
             self._updateRelations(definition)
 
-        return serialized
+    def appendAdditionalSettingDefinitions(self, additional_settings: Dict[str, Dict[str, Any]]) -> None:
```
Comment (nallath) [inline UM/Settings/DefinitionContainer.py:359]: Needs a bit of documentation. What is this doing? In what cases should one use this?

File: UM/Settings/AdditionalSettingDefinitionAppender.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+from pathlib import Path
+import os.path
+from typing import Any, Dict, List
+
+from UM.Logger import Logger
+from UM.PluginObject import PluginObject
+from UM.Settings.SettingDefinition import DefinitionPropertyType, SettingDefinition
+
+
+def prependIdToSettings(tag_type: str, tag_id: str, tag_version: str, settings: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (nallath) [inline UM/Settings/AdditionalSettingDefinitionAppender.py:14]: Consider making this a static function of the class. My main argument for it is that right now, the documentation (that is nice and extensive!) for this function takes away the attention of the documentation of the class. But the class should be the "meat" of what is going on (and thus, should probably be listed first)

**Review by nallath — CHANGES_REQUESTED** [review]
Apart from my other review comments; We need unit tests for this.

**Pull request merged**